internal/auth/oidc_authenticator.go

```diff
@@ -0,0 +1,247 @@
+	errNotAuthenticated                  = errors.New("authentication didn't succeed")
+)
+
+func newOIDCAuthenticator(cfg configauth.Authentication) (*oidcAuthenticator, error) {
+	if cfg.OIDC.Audience == "" {
+		return nil, errNoClientIDProvided
+	}
+	if cfg.OIDC.IssuerURL == "" {
+		return nil, errNoIssuerURL
+	}
+	if cfg.Attribute == "" {
+		cfg.Attribute = defaultAttribute
+	}
+
+	return &oidcAuthenticator{
+		attribute:         cfg.Attribute,
+		config:            *cfg.OIDC,
+		unaryInterceptor:  defaultUnaryInterceptor,
+		streamInterceptor: defaultStreamInterceptor,
+	}, nil
+}
+
+func (o *oidcAuthenticator) Authenticate(ctx context.Context, headers map[string][]string) (context.Context, error) {
+	authHeaders := headers[o.attribute]
+	if len(authHeaders) == 0 {
+		return ctx, errNotAuthenticated
+	}
+
+	// we only use the first header, if multiple values exist
+	parts := strings.Split(authHeaders[0], " ")
+	if len(parts) != 2 {
+		return ctx, errInvalidAuthenticationHeaderFormat
+	}
+
+	idToken, err := o.verifier.Verify(ctx, parts[1])
+	if err != nil {
+		return ctx, fmt.Errorf("failed to verify token: %w", err)
+	}
+
+	claims := map[string]interface{}{}
+	if err = idToken.Claims(&claims); err != nil {
+		// currently, this isn't a valid condition, the Verify call a few lines above
+		// will already attempt to parse the payload as a json and set it as the claims
+		// for the token. As we are using a map to hold the claims, there's no way to fail
+		// to read the claims. It could fail if we were using a custom struct. Instead of
+		// swalling the error, it's better to make this future-proof, in case the underlying
+		// code changes
+		return ctx, errFailedToObtainClaimsFromToken
+	}
+
+	sub, err := getSubjectFromClaims(claims, o.config.UsernameClaim, idToken.Subject)
+	if err != nil {
+		return ctx, fmt.Errorf("failed to get subject from claims in the token: %w", err)
+	}
+	ctx = context.WithValue(ctx, subjectKey, sub)
+
+	gr, err := getGroupsFromClaims(claims, o.config.GroupsClaim)
+	if err != nil {
+		return ctx, fmt.Errorf("failed to get groups from claims in the token: %w", err)
+	}
+	ctx = context.WithValue(ctx, groupsKey, gr)
+
+	return ctx, nil
+}
+
+func (o *oidcAuthenticator) Start(context.Context) error {
+	provider, err := getProviderForConfig(o.config)
+	if err != nil {
+		return fmt.Errorf("failed to get configuration from the auth server: %w", err)
+	}
+	o.provider = provider
+
+	o.verifier = o.provider.Verifier(&oidc.Config{
+		ClientID: o.config.Audience,
+	})
+
+	return nil
+}
+
+func (o *oidcAuthenticator) Close() error {
+	// no-op at the moment
+	// once we implement caching of the tokens we might need this
+	return nil
+}
+
+func (o *oidcAuthenticator) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	return o.unaryInterceptor(ctx, req, info, handler, o.Authenticate)
```

Comment:
@bogdandrutu just realized that moving this to `config/configauth/internal` might be easier than I thought. I'd still do it in a follow-up PR if you agree.